### PR TITLE
Split ledger-on-memory Main into multiple files

### PR DESCRIPTION
**Pull request by cocreature**
I’d like to use this in a test fixture so this PR splits up the main
and exposes an API that gives you back the resource without having to
import Main.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala [hunk 1]
```diff
@@ -29,17 +29,26 @@ import com.daml.resources.{ProgramResource, ResourceOwner}
 import scopt.OptionParser
 
 object Main {
+  private val runnerName = "In-Memory Ledger"
```
Comment (SamirTalwar-DA) [inline ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala:32]: Scala people like newlines.

Also constants are supposed to be capitalized. (I forget this a lot too.)

```suggestion
  private val RunnerName = "In-Memory Ledger"

```

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala [hunk 2]
```diff
@@ -29,17 +29,26 @@ import com.daml.resources.{ProgramResource, ResourceOwner}
...
   def main(args: Array[String]): Unit = {
     val resource = for {
       dispatcher <- dispatcherOwner
       sharedState = InMemoryState.empty
       factory = new InMemoryLedgerFactory(dispatcher, sharedState)
-      runner <- new Runner("In-Memory Ledger", factory).owner(args)
+      runner <- new Runner(runnerName, factory).owner(args)
```
Comment (SamirTalwar-DA) [inline ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala:38]:
```suggestion
      runner <- new Runner(RunnerName, factory).owner(args)
```

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala [hunk 3]
```diff
@@ -29,17 +29,26 @@ import com.daml.resources.{ProgramResource, ResourceOwner}
...
     new ProgramResource(resource).run()
   }
 
+  // Utily if you want to spin this up as a library.
+  def owner(config: Config[ExtraConfig]): ResourceOwner[Unit] =
+    for {
+      dispatcher <- dispatcherOwner
+      sharedState = InMemoryState.empty
+      factory = new InMemoryLedgerFactory(dispatcher, sharedState)
+      runner <- new Runner(runnerName, factory).owner(config)
```
Comment (SamirTalwar-DA) [inline ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala:49]:
```suggestion
      runner <- new Runner(RunnerName, factory).owner(config)
```

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala [hunk 4]
```diff
@@ -29,17 +29,26 @@ import com.daml.resources.{ProgramResource, ResourceOwner}
...
       factory = new InMemoryLedgerFactory(dispatcher, sharedState)
-      runner <- new Runner("In-Memory Ledger", factory).owner(args)
+      runner <- new Runner(runnerName, factory).owner(args)
     } yield runner
-
     new ProgramResource(resource).run()
   }
 
+  // Utily if you want to spin this up as a library.
+  def owner(config: Config[ExtraConfig]): ResourceOwner[Unit] =
```
Comment (SamirTalwar-DA) [inline ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala:44]: Can you use this in the `main` function above?

**Review by SamirTalwar-DA — COMMENTED** [review]
Seems sensible. Perhaps it's time to pull out everything that's not in `main` to their own files though. Importing from `Main` is weird.

The `RunnerName` can go into _package.scala_, and the others into their own classes/objects.

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala (hunk 4 above)
Comment (cocreature) [reply ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Main.scala:45]: Not easily, see PR description. You need the factory to parse the config so you end up having to split it up and at that point I think it just gets more convoluted.

**Comment by cocreature** [comment]
> Seems sensible. Perhaps it's time to pull out everything that's not in `main` to their own files though. Importing from `Main` is weird.

Including `owner`, if so where should that go? If you do not expose something like `owner` you have to make the pieces public which is not the case for things like `InMemoryState` atm.

**Comment by SamirTalwar-DA** [comment]
How about `Owner.apply`?

```scala
object Owner {
  def apply(config: Config[ExtraConfig]): ResourceOwner[Unit] = ...
}
```

**Comment by cocreature** [comment]
@SamirTalwar-DA made the changes you suggested

**Review by SamirTalwar-DA — APPROVED** [review]
It's so pretty. 😁

File: ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Owner.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.on.memory
+
+import com.daml.ledger.participant.state.kvutils.app.{Config, Runner}
+import com.daml.resources.ResourceOwner
+
+object Owner {
+  // Utily if you want to spin this up as a library.
```
Comment (stefanobaghino-da) [inline ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Owner.scala:10]: Utily? Did you mean useful or utility?

Comment (cocreature) [reply ledger/ledger-on-memory/src/app/scala/com/daml/ledger/on/memory/Owner.scala:10]: :facepalm:

**Pull request merged**